File: lang/php/src/fun/vac/data_structures/DoublyLinkedList.php
<?php
    namespace fun\vac\data_structures\DoublyLinkedList;

    require_once "fun/vac/data_structures/interfaces/IList.php";

    use fun\vac\data_structures\interfaces\IList;

    class Node {

        public $data;
        public $next = NULL;
        public $prev = NULL;

        public function __construct(int $element) {
            $this->data = $element;
        }
    }

    class DoublyLinkedList implements IList\IList {

        private $head = NULL;
        private $tail = NULL;
        private $length = 0;

        public function __construct() {}

        public function size(): int {
            return $this->length;
        }

        public function isEmpty(): bool {
            return $this->length == 0;
        }

        public function get(int $index): int {
            if ($index < 0 || $index >= $this->length) {
                throw new RuntimeException("[PANIC - IndexOutOfBounds]");
            }

            $cursor = NULL;

            if ($index < $this->length >> 1) {
                $cursor = $this->head;
                for ($i = 0; $i < $index; $i++) {
                    $cursor = $cursor->next;
                }
            } else {
                $cursor = $this->tail;
                for ($i = $this->length - 1; $i > $index; $i--) {
                    $cursor = $cursor->prev;
                }
            }

            return $cursor->data;
        }

        public function set(int $index, int $element): void {
            if ($index < 0 || $index >= $this->length) {
                throw new RuntimeException("[PANIC - IndexOutOfBounds]");
            }

            $cursor = NULL;

            if ($index < $this->length >> 1) {
                $cursor = $this->head;
                for ($i = 0; $i < $index; $i++) {
                    $cursor = $cursor->next;
                }
            } else {
                $cursor = $this->tail;
                for ($i = $this->length - 1; $i > $index; $i--) {
                    $cursor = $cursor->prev;
                }
            }

            $cursor->data = $element;
        }

        public function insert(int $index, int $element): void {
            if ($index < 0 || $index > $this->length) {
                throw new RuntimeException("[PANIC - IndexOutOfBounds]");
            }

            $node = new Node($element);

            switch ($index) {
            case 0:
                if ($this->length != 0) {
                    $node->next = $this->head;
                    $this->head->prev = $node;
                } else {
                    $this->tail = $node;
                }
                $this->head = $node;
                break;
            case $this->length:
                $node->prev = $this->tail;
                $this->tail->next = $node;
                $this->tail = $node;
                break;
            default:
                $cursor = NULL;
                if ($index < $this->length >> 1) {
                    $cursor = $this->head;
                    for ($i = 0; $i < $index; $i++) {
                        $cursor = $cursor->next;
                    }
                } else {
                    $cursor = $this->tail;
                    for ($i = $this->length - 1; $i > $index; $i--) {
                        $cursor = $cursor->prev;
                    }
                }
                $node->next = $cursor;
                $node->prev = $cursor->prev;
                $cursor->prev->next = $node;
                $cursor->prev = $node;
            }

            $this->length++;
        }

        public function remove(int $index): void {
            if ($index < 0 || $index >= $this->length) {
                throw new RuntimeException("[PANIC - IndexOutOfBounds]");
            }

            $target = NULL;

            switch ($index) {
            case 0:
                $target = $this->head;
                if ($this->length == 1) {
                    $this->head = NULL;
                    $this->tail = NULL;
                } else {
                    $target->next->prev = NULL;
                    $this->head = $target->next;
                }
                break;
            case $this->length - 1:
                $target = $this->tail;
                $target->prev->next = NULL;
                $this->tail = $target->prev;
                break;
            default:
                if ($index < $this->length >> 1) {
                    $target = $this->head;
                    for ($i = 0; $i < $index; $i++) {
                        $target = $target->next;
                    }
                } else {
                    $target = $this->tail;
                    for ($i = $this->length - 1; $i > $index; $i--) {
                        $target = $target->prev;
                    }
                }
                $target->prev->next = $target->next;
                $target->next->prev = $target->prev;
            }

            $target->data = NULL;

            $this->length--;
        }

        public function front(): int {
            return $this->get(0);
        }

        public function back(): int {
            return $this->get($this->length - 1);
        }

        public function prepend(int $element): void {
            $this->insert(0, $element);
        }

        public function append(int $element): void {
            $this->insert($this->length, $element);
        }

        public function poll(): void {
            $this->remove(0);
        }

        public function eject(): void {
            $this->remove($this->length - 1);
        }
    }
?>

File: lang/java/src/fun/vac/data_structures/DoublyLinkedList.java
package fun.vac.data_structures;

import fun.vac.data_structures.interfaces.IList;

public class DoublyLinkedList<E> implements IList<E> {

    private static class Node<E> {

        public E data;
        public Node<E> next = null;
        public Node<E> prev = null;

        public Node(E element) {
            data = element;
        }
    }

    private Node<E> head = null;
    private Node<E> tail = null;
    private int length = 0;

    public DoublyLinkedList() {}

    @Override
    public int size() {
        return length;
    }

    @Override
    public boolean isEmpty() {
        return length == 0;
    }

    @Override
    public E get(int index) {
        if (index < 0 || index >= length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        Node<E> cursor;

        if (index < length >>> 1) {
            cursor = head;
            for (int i = 0; i < index; i++) {
                cursor = cursor.next;
            }
        } else {
            cursor = tail;
            for (int i = length - 1; i > index; i--) {
                cursor = cursor.prev;
            }
        }

        return cursor.data;
    }

    @Override
    public void set(int index, E element) {
        if (index < 0 || index >= length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        Node<E> cursor;

        if (index < length >>> 1) {
            cursor = head;
            for (int i = 0; i < index; i++) {
                cursor = cursor.next;
            }
        } else {
            cursor = tail;
            for (int i = length - 1; i > index; i--) {
                cursor = cursor.prev;
            }
        }

        cursor.data = element;
    }

    @Override
    @SuppressWarnings("unchecked")
    public void insert(int index, E element) {
        if (index < 0 || index > length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        Node<E> node = new Node<>(element);

        if (index == 0) {
            if (length != 0) {
                node.next = head;
                head.prev = node;
            } else {
                tail = node;
            }
            head = node;
        } else if (index == length) {
            node.prev = tail;
            tail.next = node;
            tail = node;
        } else {
            Node<E> cursor;
            if (index < length >>> 1) {
                cursor = head;
                for (int i = 0; i < index; i++) {
                    cursor = cursor.next;
                }
            } else {
                cursor = tail;
                for (int i = length - 1; i > index; i--) {
                    cursor = cursor.prev;
                }
            }
            node.next = cursor;
            node.prev = cursor.prev;
            cursor.prev.next = node;
            cursor.prev = node;
        }

        length++;
    }

    @Override
    public void remove(int index) {
        if (index < 0 || index >= length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        Node<E> target;

        if (index == 0) {
            target = head;
            if (length == 1) {
                head = null;
                tail = null;
            } else {
                target.next.prev = null;
                head = target.next;
            }
        } else if (index == length - 1) {
            target = tail;
            target.prev.next = null;
            tail = target.prev;
        } else {
            if (index < length >>> 1) {
                target = head;
                for (int i = 0; i < index; i++) {
                    target = target.next;
                }
            } else {
                target = tail;
                for (int i = length - 1; i > index; i--) {
                    target = target.prev;
                }
            }
            target.prev.next = target.next;
            target.next.prev = target.prev;
        }

        target.data = null;

        length--;
    }

    @Override
    public E front() {
        return get(0);
    }

    @Override
    public E back() {
        return get(length - 1);
    }

    @Override
    public void prepend(E element) {
        insert(0, element);
    }

    @Override
    public void append(E element) {
        insert(length, element);
    }

    @Override
    public void poll() {
        remove(0);
    }

    @Override
    public void eject() {
        remove(length - 1);
    }
}
